db/batch.go

```diff
@@ -158,6 +156,37 @@ func (b *baseKVStoreBatch) Entry(index int) (*writeInfo, error) {
 	return &b.writeQueue[index], nil
 }
 
+func (b *baseKVStoreBatch) Digest() hash.Hash256 {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+	// 1. This could be improved by being processed in parallel
+	// 2. Digest could be replaced by merkle root if we need proof
+	bytes := make([]byte, 0)
+	for i := range b.writeQueue {
+		wi := &b.writeQueue[i]
+		bytes = append(bytes, wi.serialize()...)
+	}
+	return hash.Hash256b(bytes)
+}
+
+// ExcludeEntries returns copy of batch with certain entries excluded
+func (b *baseKVStoreBatch) ExcludeEntries(ns string, writeType int32) KVStoreBatch {
```

Comment:
this is to remove entries marked for future pruning in the batch
so that Digest() remains the same as usual to satisfy validateBlock()